File: output/markdown_formatter.py
"""Formats OCR + extraction results into a Markdown document."""


class MarkdownFormatter:
    """Produces a Markdown representation of one processed document."""

    def format(
        self,
        extractions: list[dict],
        source_filename: str,
        total_pages: int,
        ocr_text: str | None = None,
        cost_info: dict | None = None,
    ) -> str:
        lines: list[str] = []

        lines.append(f"# Analisi Documento: {source_filename}")
        lines.append("")
        lines.append("---")
        lines.append("")

        # Structured extractions section (only when schema is active)
        if extractions:
            lines.append("## Estrazioni Strutturate")
            lines.append("")
            for ext in extractions:
                cls = ext.get("extraction_class", "N/D")
                text = ext.get("extraction_text", "")
                lines.append(f"- **{cls}**: {text}")
                attrs = ext.get("attributes") or {}
                for k, v in attrs.items():
                    if v is not None and str(v).strip():
                        lines.append(f"  - {k}: {v}")
            lines.append("")
            lines.append("---")
            lines.append("")

        # OCR text section
        if ocr_text:
            lines.append("## Testo OCR Originale")
            lines.append("")
            lines.append(ocr_text)
            lines.append("")

        return "\n".join(lines)

File: webapp/src/lib/markdown-formatter.ts
import type { Extraction } from "./types";

export function formatMarkdown(
  extractions: Extraction[],
  sourceFilename: string,
  totalPages: number,
  ocrText: string | null
): string {
  const lines: string[] = [];

  lines.push(`# Analisi: ${sourceFilename}`);
  lines.push("");
  lines.push(`> Pagine analizzate: ${totalPages}`);
  lines.push(`> Entita' estratte: ${extractions.length}`);
  lines.push("");

  if (extractions.length > 0) {
    lines.push("## Estrazioni");
    lines.push("");

    // Group extractions by class
    const grouped: Record<string, Extraction[]> = {};
    for (const ext of extractions) {
      const cls = ext.extractionClass;
      if (!grouped[cls]) grouped[cls] = [];
      grouped[cls].push(ext);
    }

    for (const [cls, exts] of Object.entries(grouped)) {
      lines.push(`### ${cls.toUpperCase().replace(/_/g, " ")}`);
      lines.push("");

      for (const ext of exts) {
        lines.push(`- **${ext.extractionText}**`);
        if (ext.attributes && Object.keys(ext.attributes).length > 0) {
          for (const [key, value] of Object.entries(ext.attributes)) {
            lines.push(`  - _${key}_: ${value}`);
          }
        }
      }
      lines.push("");
    }
  }

  if (ocrText) {
    lines.push("---");
    lines.push("");
    lines.push("## Testo OCR");
    lines.push("");
    lines.push(ocrText);
    lines.push("");
  }

  return lines.join("\n");
}
